src/lib.rs

```diff
@@ -82,27 +87,43 @@ impl DstService {
     }
 
     #[tracing::instrument(skip(self), level = "info")]
-    async fn stream_destination(
-        &self,
-        scheme: String,
-        name: String,
-    ) -> mpsc::Receiver<GrpcResult<Update>> {
-        let state = self.state.read().await.get(&Dst { scheme, name }).cloned();
+    async fn stream_destination(&self, dst: &Dst) -> mpsc::Receiver<GrpcResult<Update>> {
+        let state = self.state.read().await.get(dst).cloned();
         if let Some(mut state) = state {
-            tracing::info!("destination exists");
+            tracing::info!("Serving endpoints");
             let (mut tx, rx) = mpsc::channel(8);
             tokio::spawn(
                 async move {
-                    let mut prev = HashSet::new();
+                    let mut prev = HashMap::new();
+
                     while let Some(Endpoints(curr)) = state.recv().await {
-                        let added = curr
-                            .difference(&prev)
-                            .map(|addr| pb::WeightedAddr {
-                                addr: Some(addr.into()),
-                                ..Default::default()
-                            })
-                            .collect::<Vec<_>>();
-                        let removed = prev.difference(&curr).map(Into::into).collect::<Vec<_>>();
+                        let (mut added, mut removed) = (vec![], vec![]);
+
+                        for (addr, EndpointMeta { h2 }) in curr.iter() {
+                            if !prev.contains_key(addr) {
+                                let protocol_hint = if *h2 {
+                                    Some(pb::ProtocolHint {
+                                        protocol: Some(pb::protocol_hint::Protocol::H2(
+                                            pb::protocol_hint::H2::default(),
+                                        )),
+                                    })
+                                } else {
+                                    None
+                                };
+                                added.push(pb::WeightedAddr {
+                                    addr: Some(addr.into()),
+                                    protocol_hint,
+                                    ..Default::default()
+                                })
+                            }
+                        }
+
+                        for addr in prev.keys() {
+                            if !curr.contains_key(addr) {
+                                removed.push(addr.into());
+                            }
+                        }
```

Comment:
TIOLI: I think this could be
```suggestion
                        let added = curr.iter()
                            .filter(|(addr, _)| !prev.contains_key(addr))
                            .map(|(addr, EndpointMeta { h2 })| {
                                let protocol_hint = if *h2 {
                                    Some(pb::ProtocolHint {
                                        protocol: Some(pb::protocol_hint::Protocol::H2(
                                            pb::protocol_hint::H2::default(),
                                        )),
                                    })
                                } else {
                                    None
                                };
                                pb::WeightedAddr {
                                    addr: Some(addr.into()),
                                    protocol_hint,
                                    ..Default::default()
                                }
                            })
                            .collect();

                        let removed = prev.keys()
                            .filter(|addr| !curr.contains_key(addr))
                            .map(Into::into)
                            .collect();
```